pkg/parse/updater.go

```diff
@@ -115,54 +133,119 @@ func (u *updater) Update(ctx context.Context, cache *cacheForCommit) status.Mult
+	var errs status.MultiError
+	errs = status.Append(errs, u.conflictErrors())
+	errs = status.Append(errs, updateErrs)
+	return errs
+}
 
+// update performs most of the work for `Update`, making it easier to
+// consistently prepend the conflict errors.
+func (u *updater) update(ctx context.Context, cache *cacheForCommit) status.MultiError {
+	// Stop remediator workers.
+	// This prevents objects been updated in the wrong order (dependencies).
+	// Continue watching previously declared objects and updating the queue.
+	// Queued objects will be remediated when the workers are started again.
+	u.remediator.Pause()
+
+	// Update the declared resources (source of truth for the Remediator)
+	var objs []client.Object
+	if cache.resourceDeclSetUpdated {
+		// No previous validation errors.
+		// Use previously validated and sanitized unstructured objects.
+		objs = u.resources.DeclaredObjects()
+	} else {
+		objs = filesystem.AsCoreObjects(cache.objsToApply)
+		var err status.MultiError
+		objs, err = u.declare(ctx, objs)
+		if err != nil {
+			return err
+		}
+		// Only mark the declared resources as updated if there were no (non-blocking) parse errors.
+		// This ensures the update will be retried.
 		if cache.parserErrs == nil {
 			cache.resourceDeclSetUpdated = true
 		}
 	}
-	// else - there were no previous validation errors
 
 	var gvks map[schema.GroupVersionKind]struct{}
-	var applyErrs status.MultiError
 	if cache.hasApplierResult {
+		// No previous apply errors.
+		// Use previously applied GVK set.
 		gvks = cache.applierResult
-		// there were no previous applier errors
 	} else {
-		applyStart := time.Now()
-		gvks, applyErrs = u.applier.Apply(ctx, objs)
-		metrics.RecordApplyDuration(ctx, metrics.StatusTagKey(applyErrs), cache.source.commit, applyStart)
-		if applyErrs != nil {
-			klog.Warningf("Failed to apply declared resources: %v", applyErrs)
-		} else if cache.parserErrs == nil {
+		var err status.MultiError
+		gvks, err = u.apply(ctx, objs, cache.source.commit)
+		if err != nil {
+			return err
+		}
+		// Only update the applier result if there were no (non-blocking) parse errors.
+		// This ensures the update will be retried.
+		if cache.parserErrs == nil {
 			cache.setApplierResult(gvks)
 		}
 	}
 
-	// Update the Remediator's watches to start new ones and stop old ones.
+	if !cache.watchesUpdated {
+		err := u.watch(ctx, gvks)
+		if err != nil {
+			return err
+		}
+		// Only mark the watches as updated if there were no (non-blocking) parse errors.
+		// This ensures the update will be retried.
+		if cache.parserErrs == nil {
+			cache.watchesUpdated = true
+		}
+	}
+
+	// Restart remediator workers.
+	// Queue will probably include all the declared objects, but they should
+	// show no diff, unless they've been updated asynchronously.
+	// Only resume after validation & apply & watch update are successful,
+	// otherwise the objects may be updated in the wrong order (dependencies).
+	u.remediator.Resume()
```

Comment:
Is it intentional that the remediator doesn't resume if we return an error earlier on in the function? Wondering if we should `defer` the Resume call

Reply:
It's intentional. If the changes are half applied and we re-enable the remediator, it may try to apply the rest of the changes the next time any event occurs for that object. For things like Roles that have no status, that may not happen without a 3rd party, but for things like Deployments that get continuous status updates from a controller and are affected by auto-scaling, it could cause changes to be made out of order, which is the problem we're trying to mitigate by pausing.